App Crashes When Canceling SMB Connection After Delay

**Describe the Bug**  
The app crashes when canceling an SMB connection after a prolonged wait time.

**Steps to Reproduce**  
1. Click on ‘+’.  
2. Click on ‘Cloud Connection’.  
3. Select ‘SMB Connection’.  
4. Wait for an extended period.  
5. Click on the ‘Cancel’ button.  
6. Click on ‘+’ again.  
7. Click on ‘Cloud Connection’.  
8. Select ‘SMB Connection’.  
9. Click on the ‘Cancel’ button.

**Expected Behavior**  
The “Searching for devices” dialog should disappear without causing a crash.

**Video**

https://github.com/user-attachments/assets/c2ea0d04-10e7-41d3-b8c7-60286c489746



**Smartphone (please complete the following information):**
 - Device: [Google Pixel 5]
 - OS: [Android 14]
 - Rooted: [No]
 - Version: [v3.10]




## Exception
* __App Name:__ Amaze File Manager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.10
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 14 - 34
* __Device:__ redfin
* __Model:__ Pixel 5
* __Product:__ redfin
<details><summary><b>Crash log </b></summary><p>

```
io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | kotlin.UninitializedPropertyAccessException: lateinit property disposable has not been initialized
	at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
	at io.reactivex.internal.observers.DisposableLambdaObserver.dispose(DisposableLambdaObserver.java:88)
	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
	at io.reactivex.internal.observers.LambdaObserver.dispose(LambdaObserver.java:102)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog.onCreateDialog$lambda$0(SmbSearchDialog.kt:83)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog.$r8$lambda$g1A-0ZtDliiR6TAsZVli3dLgYMU(Unknown Source:0)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog$$ExternalSyntheticLambda0.onClick(Unknown Source:2)
	at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:433)
	at android.view.View.performClick(View.java:7659)
	at android.view.View.performClickInternal(View.java:7636)
	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
	at android.view.View$PerformClick.run(View.java:30156)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: kotlin.UninitializedPropertyAccessException: lateinit property disposable has not been initialized
	at com.amaze.filemanager.utils.smb.SmbDeviceScannerObservable.stop(SmbDeviceScannerObservable.kt:72)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog.onCreateDialog$lambda$2(SmbSearchDialog.kt:106)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog.$r8$lambda$e-jPftKWNVfVAKamaTIyu395MpA(Unknown Source:0)
	at com.amaze.filemanager.ui.dialogs.SmbSearchDialog$$ExternalSyntheticLambda2.run(Unknown Source:2)
	at io.reactivex.internal.observers.DisposableLambdaObserver.dispose(DisposableLambdaObserver.java:85)
	... 18 more

```
</details>
<hr>

--- a/app/src/main/java/com/amaze/filemanager/utils/smb/SmbDeviceScannerObservable.kt
+++ b/app/src/main/java/com/amaze/filemanager/utils/smb/SmbDeviceScannerObservable.kt
@@ -26,6 +26,7 @@ import com.amaze.filemanager.utils.smb.SmbDeviceScannerObservable.DiscoverDevice
 import io.reactivex.Observable
 import io.reactivex.Observer
 import io.reactivex.disposables.Disposable
+import io.reactivex.disposables.Disposables
 import io.reactivex.schedulers.Schedulers
 import java.net.InetAddress
 
@@ -50,7 +51,7 @@ class SmbDeviceScannerObservable : Observable<ComputerParcelable>() {
         fun onCancel()
     }
 
-    var discoverDeviceStrategies: Array<DiscoverDeviceStrategy> =
+    private var discoverDeviceStrategies: Array<DiscoverDeviceStrategy> =
         arrayOf(
             WsddDiscoverDeviceStrategy(),
             SameSubnetDiscoverDeviceStrategy(),
@@ -61,7 +62,7 @@ class SmbDeviceScannerObservable : Observable<ComputerParcelable>() {
 
     private lateinit var observer: Observer<in ComputerParcelable>
 
-    private lateinit var disposable: Disposable
+    private var disposable: Disposable = Disposables.empty()
 
     /**
      * Stop discovering hosts. Notify containing strategies to stop, then stop the created
@@ -70,6 +71,9 @@ class SmbDeviceScannerObservable : Observable<ComputerParcelable>() {
     fun stop() {
         if (!disposable.isDisposed) {
             disposable.dispose()
+        }
+        discoverDeviceStrategies.forEach { strategy ->
+            strategy.onCancel()
         }
         observer.onComplete()
     }

